CSharpSourceCode/Battle/StatusEffects/StatusEffectMissionLogic.cs

```diff
@@ -15,28 +16,59 @@ namespace TOW_Core.Battle.StatusEffects
 {
     public class StatusEffectMissionLogic : MissionLogic
     {
+        private Dictionary<int, Agent> _activeAgents;
+        private bool init;
+
+      
         public override void OnAgentCreated(Agent agent)
         {
+            if (_activeAgents == null&&!init)
+            {
+                _activeAgents = new Dictionary<int, Agent>(); 
+                init = true;
+            }
+            
             if (agent.IsHuman)
             {
                 StatusEffectComponent effectComponent = new StatusEffectComponent(agent);
                 agent.AddComponent(effectComponent);
+                _activeAgents?.Add(agent.Index, agent);
             }
         }
         
         public override void OnMissionTick(float dt)
         {
-            foreach(var agent in Mission.Current.Agents)
+            
+            if (!init)
+                return;
+
+            foreach (var agent in _activeAgents.ToDictionary(x =>x.Key, x=> x.Value))
             {
-                if (agent.GetComponent<StatusEffectComponent>() != null)
+                if (agent.Value.GetComponent<StatusEffectComponent>() != null)
                 {
-                    if (agent.IsActive() && agent.Health > 1f)
+                    if (agent.Value.GetComponent<StatusEffectComponent>().IsDisabled())
+                    {
+                        _activeAgents.Remove(agent.Key);
+                        continue;
+                    }
+                    if (agent.Value.IsActive() && agent.Value.Health > 1f)
                     {
-                        var comp = agent.GetComponent<StatusEffectComponent>();
+                        var comp = agent.Value.GetComponent<StatusEffectComponent>();
                         comp.OnTick(dt);
                     }
+
+                    
                 }
```

Comment:
as for me would be a bit better to write so
![image](https://user-images.githubusercontent.com/62873054/144464128-ddec6cef-0efd-46a1-b0bb-0ac92faf8bd8.png)
but it's just a suggestion

Reply:
worked on this